substrate/client/consensus/aura/src/authorities_tracker.rs

```diff
@@ -0,0 +1,180 @@
+	traits::{Block, Header, NumberFor},
+};
+
+use crate::{fetch_authorities_from_runtime, AuthorityId, CompatibilityMode};
+
+const LOG_TARGET: &str = "aura::authorities_tracker";
+
+/// AURA authorities tracker. Updates authorities based on the AURA authorities change
+/// digest in the block header.
+pub struct AuthoritiesTracker<P: Pair, B: Block, C> {
+	authorities: RwLock<ForkTree<B::Hash, NumberFor<B>, Vec<AuthorityId<P>>>>,
+	client: Arc<C>,
+}
+
+impl<P: Pair, B: Block, C> AuthoritiesTracker<P, B, C> {
+	/// Create a new `AuthoritiesTracker`.
+	pub fn new(client: Arc<C>) -> Self {
+		Self { authorities: RwLock::new(ForkTree::new()), client }
+	}
+}
+
+impl<P, B, C> AuthoritiesTracker<P, B, C>
+where
+	P: Pair,
+	B: Block,
+	C: HeaderBackend<B> + HeaderMetadata<B, Error = sp_blockchain::Error> + ProvideRuntimeApi<B>,
+	P::Public: Codec + Debug,
+	C::Api: AuraApi<B, AuthorityId<P>>,
+{
+	/// Fetch authorities from the tracker, if available. If not available, fetch from the client
+	/// and update the tracker.
+	pub fn fetch_or_update(
+		&self,
+		header: &B::Header,
+		compatibility_mode: &CompatibilityMode<NumberFor<B>>,
+	) -> Result<Vec<AuthorityId<P>>, String> {
+		let hash = header.hash();
+		let number = *header.number();
+		let parent_hash = *header.parent_hash();
+
+		// Fetch authorities from cache, if available.
+		let authorities = {
+			let is_descendent_of =
+				sc_client_api::utils::is_descendent_of(&*self.client, Some((hash, parent_hash)));
+			let authorities_cache = self.authorities.read();
+			authorities_cache
+				.find_node_where(&hash, &number, &is_descendent_of, &|_| true)
+				.map_err(|e| {
+					format!("Could not find authorities for block {hash:?} at number {number}: {e}")
+				})?
+				.map(|node| node.data.clone())
+		};
+
+		match authorities {
+			Some(authorities) => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Authorities for block {:?} at number {} found in cache",
+					hash,
+					number,
+				);
+				Ok(authorities)
+			},
+			None => {
+				// Authorities are missing from the cache. Fetch them from the runtime and cache
+				// them.
+				log::debug!(
+					target: LOG_TARGET,
+					"Authorities for block {:?} at number {} not found in cache, fetching from runtime",
+					hash,
+					number
+				);
+				let authorities = fetch_authorities_from_runtime(
+					&*self.client,
+					parent_hash,
+					number,
+					compatibility_mode,
+				)
+				.map_err(|e| format!("Could not fetch authorities at {:?}: {}", parent_hash, e))?;
+				let is_descendent_of = sc_client_api::utils::is_descendent_of(&*self.client, None);
+				let mut authorities_cache = self.authorities.write();
+				authorities_cache
+					.import(
+						parent_hash,
+						number - 1u32.into(),
+						authorities.clone(),
+						&is_descendent_of,
+					)
+					.map_err(|e| {
+						format!("Could not import authorities for block {parent_hash:?} at number {}: {e}", number - 1u32.into())
+					})?;
+				Ok(authorities)
+			},
```

Comment:
This code path should not exist in this function. We should warm up the cache at startup and fetch the authorities for all blocks from the leaves back to the finalized block.